### PR TITLE
Remove marshmallow from result/ and requirements

### PR DESCRIPTION
**Pull request by mtreinish**
### Summary

This commit finishes the process of removing marshmallow from terra. It
rebuilds the result class as bare python classes or SimpleNamespace
subclasses (for classes that allow arbitrary fields). After the result
class has been converted there is nothing using marshmallow or all the
support code in qiskit/validation. So this continues the removal process
by removing those and removing marshmallow and marshmallow-polyfield
from the requirements list.

### Details and comments

This commit depends on #4027 and #4016 and can't be merged (or tested) without
those applied.

### CODE REVIEW

**Comment by mtreinish** [comment]
This is still on hold because this will break the ibmq-provider until Qiskit/qiskit-ibmq-provider#553 is fixed. Once that's done we can safely merge this

**Comment by mtreinish** [comment]
Since marshmallow was removed in qiskit-ibmq-provider 0.7.0 this is unblocked and should be ready to merge

File: qiskit/validation/__init__.py [hunk 1]
```diff
@@ -19,25 +19,19 @@
 
 .. currentmodule:: qiskit.validation
 
-Base
-====
+JSON Schema
+===========
 
-.. autosummary::
+,, autosummary::
```
Comment (jyu00) [inline qiskit/validation/__init__.py:25]: Did you mean to change this from dots to commas?

File: qiskit/result/result.py [hunk 2]
```diff
@@ -14,19 +14,19 @@
...
-from .models import ResultSchema
+from qiskit.qobj import QobjHeader
 
 
-@bind_schema(ResultSchema)
-class Result(BaseModel):
+class Result:
     """Model for Results.
 
     Please note that this class only describes the required fields. For the
```
Comment (jyu00) [inline qiskit/result/result.py:32]: Lines 32-33 are probably no longer needed as there is no `ResultSchema` anymore.

File: qiskit/validation/__init__.py (hunk 1 above)
Comment (mtreinish) [reply qiskit/validation/__init__.py:25]: oops, nope. I'll fix that in a sec

File: qiskit/result/result.py (hunk 2 above)
Comment (mtreinish) [reply qiskit/result/result.py:32]: Yep good catch, I meant to update this

Comment (mtreinish) [reply qiskit/result/result.py]: Fixed in: https://github.com/Qiskit/qiskit-terra/pull/4030/commits/514d58fef10d2e3db29bfe4c7a71fb1ddf121c9a

File: qiskit/validation/__init__.py (hunk 1 above)
Comment (mtreinish) [reply qiskit/validation/__init__.py]: Fixed in: https://github.com/Qiskit/qiskit-terra/pull/4030/commits/514d58fef10d2e3db29bfe4c7a71fb1ddf121c9a

**Comment by mtreinish** [comment]
Honestly, I'm not sure if that's a concern for this PR. Those methods are there for backwards compatibility with marshmallow only, if I could I would have just removed them. Ideally if I were to add json support to these classes I would have done it with `to_json`/`from_json` (or maybe `dump_json`) or probably even better would just be `save` and `load` and had it actually do the serialization. In practice though I don't know if people (outside of the ibmq provider) are actually using it with json, the places I would expect it to be used don't (like ignis for saving experimental results relies on pickles and ignores them). Other (non-ibmq) providers typically actually do `from_dict` but not from loading json instead from a manually constructed dict, like:

- https://github.com/Qiskit/qiskit-honeywell-provider/blob/a08a4d12bb0493a9c40462cba1b4e23c4ea2972d/qiskit/providers/honeywell/honeywelljob.py#L303-L311
- https://github.com/Qiskit/qiskit-aqt-provider/blob/master/qiskit/providers/aqt/aqt_job.py#L90-L106 
- https://github.com/Qiskit/qiskit-terra/blob/master/qiskit/providers/basicaer/qasm_simulator.py#L417-L427

mostly because the marshmallow based classes had poor documentation and if you were going to construct a Results object using the constructors it relied on three layers of nested class construction  so `from_dict`was much easier.

**Comment by jyu00** [comment]
The `to_dict` and `from_dict` don't really do JSON (de)serialization, as they don't convert non-JSON-serializable objects (e.g. `complex`) to JSON serializable. They really just convert an object to its dictionary representation.

**Pull request merged**